### DISCUSSION TITLE
BUilding mfem on windows

### DISCUSSION THREAD

**Discussion by starkekr** [mfem/mfem#4338] [General]
What is the preffered way to build mfem on windows?

I have tried to use msys2. To do this I have tried to install all the dependencies of the mfem package located here:
https://packages.msys2.org/package/mingw-w64-ucrt-x86_64-mfem?repo=ucrt64
Then when I go to build mfem, I get the following error:

![image](https://github.com/mfem/mfem/assets/118568957/ef1f2361-9d5d-42ac-b79c-a27533a7cdcd)

I have also tried following the documentation online but I get stuck building both Hypre and Metis in the msys2 terminal. Are there compiler version constraints that I need to adhere to?

**Comment by v-dobrev** [comment]
Did you try installing the binary package from MSYS2? If that works for you, the examples can be built to link with that installation. For example, you can try building `ex0p` with something like:
```console
make CONFIG_MK=<path-to-mfem-install-dir>/share/mfem/config.mk ex0p
```
The `config.mk` file found in the `.tar.zst` file from the page you linked above (https://packages.msys2.org/package/mingw-w64-ucrt-x86_64-mfem?repo=ucrt64) has some paths that you may need to adjust. For example, it sets `MFEM_CXX = D:/a/msys64/ucrt64/bin/g++.exe` and the paths to TPLs in the variable `MFEM_EXT_LIBS` use prefix `D:/a/msys64/ucrt64`, e.g. hypre's library is given as `D:/a/msys64/ucrt64/lib/libHYPRE.dll.a`. If install the MSYS2 MFEM package in a different location, these paths will need to be updated.

If you prefer to build MFEM yourself, the first steps will be to build the hypre and metis dependencies (if you want MPI-enabled build; in serial, MFEM can be built without any dependencies). Can you post the build commands you used for hypre and metis, and what are the errors you got?

**Reply by starkekr** [reply]
I would like to build mfem myself. I was able to build hypre with no problems using cmake. With metis though, I get the following error after running cmake:

![image](https://github.com/mfem/mfem/assets/118568957/0dfa90cf-53ce-4a55-824c-ca26cec25314)

**Comment by v-dobrev** [comment]
We have a set of patches for Metis 5.1.0 that we apply in CI testing on Windows with vcpkg -- see `config/vcpkg/ports/metis-mfem/*.patch` in the MFEM source directory -- you can try to apply them before building Metis. I'm not sure if they are all needed when building under MSYS2 but hopefully they'll fix the build issues.

MSYS2 also has a set of patches for Metis that you can try, see https://github.com/msys2/MINGW-packages/tree/master/mingw-w64-metis -- these are probably more appropriate for building under MSYS2.

**Reply by starkekr** [reply]
I was able to get it working using a patch for metis 4.0.03, Thanks for the help!
